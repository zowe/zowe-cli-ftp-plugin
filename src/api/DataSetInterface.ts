/*
 * This program and the accompanying materials are made available under the terms of the
 * Eclipse Public License v2.0 which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-v20.html
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Copyright Contributors to the Zowe Project.
 *
 */

import { IFTPProgressHandler } from "./IFTPProgressHandler";

/**
 * Bytes per track.
 */
export const TRACK = 56664;

export interface IDownloadDataSetOption {
    /**
     * Optional local file that the dowloaded dataset will be saved to.
     */
    localFile?: string;

    /**
     * Optional repsonse handler for progressing messages.
     */
    progress?: IFTPProgressHandler;

    /**
     * TRANSFER_TYPE_ASCII, TRANSFER_TYPE_BIANRY, TRANSFER_TYPE_ASCII_RDW, or TRANSFER_TYPE_BINARY_RDW defined in CoreUtils.
     */
    transferType?: string;

    /**
    * Optional encoding for download and upload of z/OS data set
    */
    encoding?: string;
}

export interface IUploadDataSetOption {
    /**
     * Optional content that will be uploaded, if `localFile` is not specified.
     */
    content?: Buffer | string;

    /**
     * Data control block (DCB). Refer to https://github.com/IBM/zos-node-accessor/tree/1.0.7#allocate
     */
    dcb?: string;

    /**
     * Optional local file that will be uploaded.
     */
    localFile?: string;

    /**
     * TRANSFER_TYPE_ASCII or TRANSFER_TYPE_BIANRY defined in CoreUtils.
     */
    transferType?: string;

    /**
    * Optional encoding for download and upload of z/OS data set
    */
    encoding?: string;
}

export interface IAllocateDataSetOption {

    /**
     * Data control block (DCB). Refer to https://github.com/IBM/zos-node-accessor/tree/1.0.7#allocate
     */
    dcb?: string;
}

/**
 * Copy Dataset Options
 */
export interface ICopyDataSetOptions {
    /**
     * Required: Name of the SOURCE dataset
     */
    fromDsn: string;

    /**
     * Required: Name of the TARGET dataset
     */
    toDsn: string;

    /**
     * Optional: Progress indicator object/task
     */
    progress?: IFTPProgressHandler;

    /**
     * Optional: Indicator to force a replacement
     */
    replace?: boolean;
}

/**
 * Detailed allocation options for a dataset
 */
export interface IDataSetDetailedAllocationOptions {
    /**
     * Dataset volume serial
     */
    volume?: string; // Not supported by connection.allocateDataset

    /**
     * Dataset record format
     */
    recfm?: string;

    /**
     * Dataset block size
     */
    BLOCKSIze?: string; // Strange mapping

    /**
     * Dataset record length
     */
    lrecl?: string;

    /**
     * Dataset organization (PS vs PO)
     */
    dsorg?: string;

    /**
     * The following values are not supported by the allocateDataset method of zos-node-accessor@v1
     * However, they are return as allocation details for a dataset
     */
    // BLocks: ds.??? // Strage mapping + Not returned by connection.listDataset
    // CYlinders: ds.??? // Strage mapping + Not returned by connection.listDataset
    // TRacks: ds.??? // Strage mapping + Not returned by connection.listDataset
    // Directory: ds.??? // Strage mapping + Not returned by connection.listDataset
    // PRImary: ds.??? // Strage mapping + Not returned by connection.listDataset
    // SECondary: ds.??? // Strage mapping + Not returned by connection.listDataset
}

// When DataSetUtilsV2 for zos-node-accessor v2 is ready, alias DataSetUtilsV2 to DataSetUtils.
// export { DataSetUtils as DataSetUtils };
