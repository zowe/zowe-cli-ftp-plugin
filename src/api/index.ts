/*
 * This program and the accompanying materials are made available under the terms of the
 * Eclipse Public License v2.0 which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-v20.html
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Copyright Contributors to the Zowe Project.
 *
 */

export * as zosNodeAccessor from "zos-node-accessor";
export { TransferMode } from "zos-node-accessor";

export * from "./doc";

export * from "./CoreUtils";
export * from "./FTPConfig";
export * from "./JobUtils";
export * from "./DataSetUtils";
export * from "./UssUtils";
