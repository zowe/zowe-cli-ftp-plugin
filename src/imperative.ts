/*
 * This program and the accompanying materials are made available under the terms of the
 * Eclipse Public License v2.0 which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-v20.html
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Copyright Contributors to the Zowe Project.
 *
 */

import { IImperativeConfig } from "@brightside/imperative";
import { FTPConfig } from "./api/FTPConfig";

const config: IImperativeConfig = {
    commandModuleGlobs: ["**/cli/*/*.definition!(.d).*s"],
    rootCommandDescription: "Data set and job functionality via FTP. This functionality " +
    "uses the open source zos-node-accessor package from IBM. Commands under this group " +
    "require you to create a zftp profile before using them.\n" +
    "If you find this functionality useful, please consider setting up z/OSMF on " +
    "your system to get improved stability and speed and more features (for example, issuing TSO and console commands) by using " +
    "core Zowe CLI.",
    productDisplayName: "z/OS FTP Plugin",
    pluginAliases: ["zftp"],
    pluginSummary: "z/OS Files and jobs via FTP",
    pluginHealthCheck: "./lib/healthCheck.handler",
    name: "zos-ftp",
    profiles: [
        {
            type: "zftp",
            createProfileExamples: [{
                options: "myprofile -u ibmuser -p ibmp4ss -H sys123",
                description: "Create a zftp profile called 'myprofile' with default settings (port" +
                ", timeout, etc.) to connect with the host system 123."
            }],
            schema: {
                type: "object",
                title: "Configuration profile for z/OS FTP",
                description: "Configuration profile for z/OS FTP",
                properties: {
                    host: {
                        type: "string",
                        optionDefinition: FTPConfig.OPTION_HOST
                    },
                    port: {
                        type: "number",
                        optionDefinition: FTPConfig.OPTION_PORT
                    },
                    user: {
                        type: "string",
                        optionDefinition: FTPConfig.OPTION_USER,
                        secure: true
                    },
                    password: {
                        type: "string",
                        optionDefinition: FTPConfig.OPTION_PASSWORD,
                        secure: true
                    },
                    secureFtp: {
                        type: "string",
                        optionDefinition: FTPConfig.OPTION_SECURE_FTP
                    },
                    rejectUnauthorized: {
                        type: ["boolean", "null"],
                        optionDefinition: FTPConfig.OPTION_REJECT_UNAUTHORIZED
                    },
                    servername: {
                        type: ["string", "null"],
                        optionDefinition: FTPConfig.OPTION_SERVER_NAME
                    },
                    connectionTimeout: {
                        type: "number",
                        optionDefinition: FTPConfig.OPTION_CONNECTION_TIMEOUT
                    },
                },
                optional: ["host", "port", "user", "password"],
            },
        },
    ]
};

export = config;
