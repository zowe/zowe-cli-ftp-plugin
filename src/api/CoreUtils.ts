/*
 * This program and the accompanying materials are made available under the terms of the
 * Eclipse Public License v2.0 which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-v20.html
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Copyright Contributors to the Zowe Project.
 *
 */

import { ICommandProfileTypeConfiguration, IImperativeError, Logger } from "@zowe/imperative";
import * as stream from "stream";
import { TransferMode } from "zos-node-accessor";

export class CoreUtils {

    /**
     * Sleep for a number of milliseconds
     * @param ms - how many miliseconds to sleep or (at least)
     * @returns promise that resolves when the sleep is complete (or after, not guaranteed)
     */
    public static async sleep(ms: number) {
        this.log.debug("Sleeping for %s milliseconds...", ms);
        return new Promise((resolve) => setTimeout(resolve, ms));
    }

    /**
     * Make sure newlines in a string have carriage returns
     * This is required to prevent errors when transferring
     * in ascii mode
     * @param  original - the original string which you would like to ensure has \r before \n
     * @returns a string where every \n has a \r before it
     */
    public static addCarriageReturns(original: string): string {
        this.log.debug("Adding carriage returns to string of length %s", original.length);
        return original.replace(/\r?\n/g, "\r\n");
    }

    /**
     * Read the complete contents of stdin
     * @param stream - A custom stream to read stdin data from
     * @returns contents piped in to stdin
     */
    public static async readStdin(stream?: stream.Readable): Promise<Buffer> {
        return new Promise<Buffer>((resolve, reject) => {
            let stdinContent: Buffer = Buffer.from([]);
            stream = stream || process.stdin;
            stream.resume();

            stream.on("data", (chunk: Buffer) => {
                this.log.trace(`Read data from stdin: ${chunk.toString()}`);
                this.log.debug(`Read ${chunk.length} bytes of data from stdin`);
                stdinContent = Buffer.concat([stdinContent, chunk]);
            });

            stream.once("end", () => {
                this.log.info("Finished reading stdin");
                resolve(stdinContent);
            });

            stream.on("error", (error: Error) => {
                const stdinReadError: IImperativeError = {
                    msg: "Error encountered while reading from stdin",
                    causeErrors: error,
                    additionalDetails: error?.message
                };
                reject(stdinReadError);
            });
        });
    }

    public static async getProfileSchema(): Promise<ICommandProfileTypeConfiguration[]> {
        const ftpProfile = await require("../imperative").profiles as ICommandProfileTypeConfiguration[];
        return ftpProfile;
    }

    protected static get log(): Logger {
        return Logger.getAppLogger();
    }

    public static getBinaryTransferModeOrDefault(isBinary: boolean, rdw = false): TransferMode {
        return (isBinary ?
            (rdw ? TransferMode.BINARY_RDW : TransferMode.BINARY) :
            (rdw ? TransferMode.ASCII_RDW : TransferMode.ASCII)
        );
    }
}
