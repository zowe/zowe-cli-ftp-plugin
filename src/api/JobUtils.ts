/*
 * This program and the accompanying materials are made available under the terms of the
 * Eclipse Public License v2.0 which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-v20.html
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Copyright Contributors to the Zowe Project.
 *
 */

import { Logger } from "@zowe/imperative";
import { IGetSpoolFileOption, IJob, IJobStatus, IListJobOption, ISpoolFile } from "./JobInterface";

export class JobUtils {

    /**
     * Lists jobs with the job name prefix.
     *
     * @param connection - zos-node-accessor connection
     * @param prefix - job name prefix
     * @param option - list job option
     * @returns job entries
     */
    public static async listJobs(connection: any, prefix: string, option?: IListJobOption): Promise<IJob[]> {
        const accessorOption: any = {
            jobName: prefix || "*",
        };
        let debugMessage = `Listing jobs that match prefix ${prefix}`;
        if (option && option.owner) {
            accessorOption.owner = option.owner;
            debugMessage += ` and are owned by ${accessorOption.owner}`;
        }
        if(option && option.status) {
            accessorOption.status = option.status;
            debugMessage += ` and status is ${accessorOption.status}`;
        }
        this.log.debug(debugMessage);

        const jobs = await connection.listJobs(accessorOption);
        this.log.debug("List returned %d jobs", jobs.length);
        const filteredJobs = JobUtils.parseJobDetails(jobs);
        return filteredJobs;
    }

    /**
     * Deletes the job with the job id.
     *
     * @param connection - zos-node-accessor connection
     * @param jobId - job id
     */
    public static async deleteJob(connection: any, jobId: string): Promise<void> {
        this.log.debug("Deleting job with job id '%s'", jobId);
        await connection.deleteJob(jobId);
    }

    /**
     * Return the job spool files content of the specified job id and spool file id.
     *
     * @param connection - zos-node-accessor connection
     * @param option - option
     * @returns spool file content
     */
    public static async getSpoolFileContent(connection: any, option: IGetSpoolFileOption): Promise<Buffer> {
        return connection.getJobLog(option);
    }

    /**
     * Return the all job spool files content of the specified job id.
     *
     * @param connection - zos-node-accessor connection
     * @param jobId - job ID
     * @returns spool files with content
     */
    public static async getSpoolFiles(connection: any, jobId: string): Promise<ISpoolFile[]> {
        const jobDetails = (await JobUtils.findJobByID(connection, jobId));
        const fullSpoolFiles: ISpoolFile[] = [];
        if (jobDetails.spoolFiles) {
            for (const spoolFileToDownload of jobDetails.spoolFiles) {
                this.log.debug("Requesting spool files for job %s(%s) spool file ID %d", 
                    jobDetails.jobname, jobDetails.jobid, spoolFileToDownload.id);
                const option = {
                    jobName: jobDetails.jobname,
                    jobId: jobDetails.jobid,
                    owner: "*",
                    fileId: spoolFileToDownload.id
                };
                const spoolFile = await JobUtils.getSpoolFileContent(connection, option);
                spoolFileToDownload.contents = spoolFile;
                fullSpoolFiles.push(spoolFileToDownload);
            }
        }
        return fullSpoolFiles;
    }

    /**
     * Submits job with the specified JCL.
     *
     * @param connection - zos-node-accessor connection
     * @param jcl - jcl
     * @returns job id
     */
    public static async submitJob(connection: any, jcl: string): Promise<string> {
        return connection.submitJCL(jcl);
    }

    /**
     * Submits job with the specified JCL dataset.
     *
     * @param connection - zos-node-accessor connection
     * @param dsn - fully-qualified JCL dataset name without quotes
     * @returns job id
     */
    public static async submitJobFromDataset(connection: any, dsn: string): Promise<string> {
        const dsContent = (await connection.getDataset("'" + dsn + "'")).toString();
        this.log.debug("Downloaded data set '%s'. Submitting...", dsn);
        return JobUtils.submitJob(connection, dsContent);
    }

    /**
     * Identify a job by job ID
     * Warning: slow since it lists all jobs
     * @param  jobId - the job ID of the job you want to identify -
     *                         note: you can't use the abbreviated version like j123. it must be the full job ID
     * @param connection - connection to zos-node-accessor
     */
    public static async findJobByID(connection: any, jobId: string): Promise<IJobStatus> {
        this.log.debug("Attempting to locate job by job ID %s", jobId);
        const jobStatus = await connection.getJobStatus({jobId: jobId.toUpperCase(), owner: "*"});
        if (jobStatus.retcode) {
            // zos-node-accessor returns 'RC 0000', which need be converted to 'CC 0000'.
            jobStatus.retcode = jobStatus.retcode.replace(/^RC /, "CC ");
        }
        return jobStatus;
    }

    public static parseJobDetails(jobs: string[]): IJob[] {
        if (jobs.length > 1) {
            jobs = jobs.slice(1);
        }
        return jobs.map((job: string) => {
            // object looks like:
            // JOBNAME, JOBID, OWNER, STATUS, CLASS
            // turn the object into a similar format to that returned by
            // z/osmf so that users who use the list ds command in main
            // zowe can use the same filtering options
            const fields = job.split(/\s+/g);

            const jobNameIndex = 0;
            const jobIdIndex = 1;
            const ownerIndex = 2;
            const statusIndex = 3;
            const classIndex = 4;
            return {
                jobname: fields[jobNameIndex],
                jobid: fields[jobIdIndex],
                owner: fields[ownerIndex],
                status: fields[statusIndex],
                class: fields[classIndex],
                originalFtpResult: job
            };
        });
    }

    private static get log(): Logger {
        return Logger.getAppLogger();
    }
}
