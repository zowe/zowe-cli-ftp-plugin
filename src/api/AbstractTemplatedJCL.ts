/*
 * This program and the accompanying materials are made available under the terms of the
 * Eclipse Public License v2.0 which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-v20.html
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Copyright Contributors to the Zowe Project.
 *
 */

import { IO, Logger, TextUtils } from "@zowe/imperative";
import { CoreUtils } from "./CoreUtils";

export abstract class AbstractTemplatedJCL {

    /**
     * Default job template for submitting your JCL
     */
    public abstract readonly DEFAULT_TEMPLATE: string;

    /**
     *
     * @param jobCardFile - file containing a job card which will be prepended to the job
     * @param  substitutionValues -
     * @param  templateFile - override the default JCL template with a file
     */
    public getJcl(jobCardFile: string, substitutionValues: {[key: string]: string}, templateFile?: string) {
        let jcl = IO.readFileSync(jobCardFile).toString();
        if (!/\r?\nZ/.test(jcl)) {
            // add a newline to the end of the jcl if there is none
            jcl += "\r\n";
        }
        if (templateFile != null) {
            // we
            jcl += IO.readFileSync(templateFile).toString();
        } else {
            jcl += this.DEFAULT_TEMPLATE;
        }
        return TextUtils.renderWithMustache(CoreUtils.addCarriageReturns(jcl), substitutionValues);
    }

    protected get log(): Logger {
        return Logger.getAppLogger();
    }
}
