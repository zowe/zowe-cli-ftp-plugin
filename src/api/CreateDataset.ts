/*
 * This program and the accompanying materials are made available under the terms of the
 * Eclipse Public License v2.0 which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-v20.html
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Copyright Contributors to the Zowe Project.
 *
 */

import { CreateDataSetTypeEnum, CreateDefaults, ICreateDataSetOptions, IZosFilesResponse, ZosFilesMessages } from "@zowe/cli";
import { ImperativeError, ImperativeExpect, TextUtils } from "@zowe/imperative";
import { JobUtils } from "./JobUtils";
import { CoreUtils } from "./CoreUtils";
import { AbstractTemplatedJCL } from "./AbstractTemplatedJCL";

export class CreateDataset extends AbstractTemplatedJCL {

    /**
     * Create a data set
     * @param connection                 - zos-node-accessor connection
     * @param  cmdType               - The type of data set we are going to create
     * @param dataSetName                          - the name of the data set to create
     * @param  [options={}] - additional options for the creation of the data set
     */
    public static async create(connection: any, cmdType: CreateDataSetTypeEnum,
        dataSetName: string,
        jobCardFile: string,
        options ?: Partial<ICreateDataSetOptions>):
        Promise<IZosFilesResponse> {
        let validCmdType = true;

        // Removes undefined properties
        let tempOptions = options != null ? JSON.parse(JSON.stringify(options)) : {};
        const secondarySpecified = tempOptions.secondary != null;

        // Required
        ImperativeExpect.toNotBeNullOrUndefined(cmdType, ZosFilesMessages.missingDatasetType.message);

        // Required
        ImperativeExpect.toNotBeNullOrUndefined(dataSetName, ZosFilesMessages.missingDatasetName.message);

        switch (cmdType) {
            case CreateDataSetTypeEnum.DATA_SET_PARTITIONED:
                tempOptions = {...CreateDefaults.DATA_SET.PARTITIONED, ...tempOptions};
                break;
            case CreateDataSetTypeEnum.DATA_SET_SEQUENTIAL:
                tempOptions = {...CreateDefaults.DATA_SET.SEQUENTIAL, ...tempOptions};
                break;
            case CreateDataSetTypeEnum.DATA_SET_BINARY:
                tempOptions = {...CreateDefaults.DATA_SET.BINARY, ...tempOptions};
                break;
            case CreateDataSetTypeEnum.DATA_SET_C:
                tempOptions = {...CreateDefaults.DATA_SET.C, ...tempOptions};
                break;
            case CreateDataSetTypeEnum.DATA_SET_CLASSIC:
                tempOptions = {...CreateDefaults.DATA_SET.CLASSIC, ...tempOptions};
                break;
            default:
                validCmdType = false;
                break;
        }

        if (!validCmdType) {
            throw new ImperativeError({msg: ZosFilesMessages.unsupportedDatasetType.message});
        } else {
            // Handle the size option
            if (tempOptions.size != null) {
                const tAlcunit = tempOptions.size.toString().match(/[a-zA-Z]+/g);
                if (tAlcunit != null) {
                    tempOptions.alcunit = tAlcunit.join("").toUpperCase();
                }

                const tPrimary = tempOptions.size.toString().match(/[0-9]+/g);
                if (tPrimary != null) {
                    tempOptions.primary = +(tPrimary.join(""));

                    if (!secondarySpecified) {
                        const TEN_PERCENT = 0.10;
                        tempOptions.secondary = Math.round(tempOptions.primary * TEN_PERCENT);
                    }
                }

                delete tempOptions.size;
            }

            let response = "";
            // Handle the print attributes option
            if (tempOptions.printAttributes != null) {
                if (tempOptions.printAttributes) {
                    delete tempOptions.printAttributes;
                    response = TextUtils.prettyJson(tempOptions);
                } else {
                    delete tempOptions.printAttributes;
                }
            }

            response = await new CreateDataset().createViaFTP(connection, dataSetName, tempOptions, jobCardFile);
            if (response.indexOf("RC=0000") >= 0) {
                return {
                    success: true,
                    commandResponse: response + "\n" + ZosFilesMessages.dataSetCreatedSuccessfully.message,
                }
            } else {
                return {
                    success: false,
                    commandResponse: response + "\nFailed to create data set",
                }
            }
        }
    }


    public DEFAULT_TEMPLATE = "//CREATE EXEC PGM=IEFBR14\n" +
        "//{{createDD}}";

    private async createViaFTP(connection: any, dataSetName: string,
        options: ICreateDataSetOptions,
        jobCardFile: string,
        overrideTemplateFile ?: string) {
        this.log.debug("Building jcl to create data set %s", dataSetName);

        let createDD: string = "NEWDATA DD ";
        const createDDValues = ["DSN=" + dataSetName.toUpperCase(), "DISP=(NEW,CATLG)"];
        const maxDDLineLength = 65;
        let currentLineLength = createDD.length;
        const continueLine = "//   ";

        // todo: make sure all options are used here
        if (options.alcunit) {
            let spaceParams = "SPACE=(";
            spaceParams += options.alcunit.toUpperCase() + ",(";
            spaceParams += options.primary;
            if (options.secondary != null) {
                spaceParams += "," + options.secondary;
            }
            spaceParams += "))";

            createDDValues.push(spaceParams);
            if (options.volser != null) {
                createDDValues.push("VOLSER=" + options.volser.toUpperCase());
            }
        }
        if (options.lrecl != null) {
            createDDValues.push("LRECL=" + options.lrecl);
        }
        if (options.recfm != null) {
            createDDValues.push("RECFM=" + options.recfm.toUpperCase());
        }
        if (options.blksize != null) {
            createDDValues.push("BLKSIZE=" + options.blksize);
        }
        if (options.unit != null) {
            createDDValues.push("UNIT=" + options.unit.toUpperCase());
        }

        if (options.dsorg != null) {
            createDDValues.push("DSORG=" + options.dsorg.toUpperCase());
        }

        /**
         * Build the DD definition for the job by looping through the values we've built
         */
        for (let i = 0; i < createDDValues.length; i++) {
            const value = createDDValues[i];
            if (value.length + currentLineLength > maxDDLineLength) {
                createDD += "\n" + continueLine;
                currentLineLength = continueLine.length;
            }

            createDD += value;
            if (i < createDDValues.length - 1) {
                // if this isn't the last value, add a comma to continue the line
                createDD += ",";
                currentLineLength += 1;
            }
            currentLineLength += value.length;
        }

        const jcl = this.getJcl(jobCardFile, {createDD}, overrideTemplateFile);
        this.log.debug("Creating data set %s via JCL:\n %s", dataSetName, jcl);
        const jobid = await connection.submitJCL(jcl);
        const jobWait = 500;
        await CoreUtils.sleep(jobWait); // have to wait for the job to show up in the list
        const jobDetails = await JobUtils.findJobByID(connection, jobid);
        return "Job details: " + JSON.stringify(jobDetails, null, 2);
    }

}
