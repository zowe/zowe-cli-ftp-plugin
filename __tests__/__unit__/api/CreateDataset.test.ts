/*
 * This program and the accompanying materials are made available under the terms of the
 * Eclipse Public License v2.0 which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-v20.html
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Copyright Contributors to the Zowe Project.
 *
 */

import { ImperativeError, IO } from "@zowe/imperative";
import { CreateDataSetTypeEnum } from "@zowe/zos-files-for-zowe-sdk";
import { CreateDataset } from "../../../src/api/CreateDataset";
import { JobUtils } from "../../../src/api/JobUtils";

describe("CreateDataset", () => {
    afterEach(() => {
        jest.restoreAllMocks();
    });

    it("should fail if requested to create a dataset of an unsuported type", async () => {
        const connection = {};
        let caughtError: ImperativeError;
        try {
            await CreateDataset.create(connection, -1, "dsname", "jobCard", {});
            throw "FAILURE";
        } catch(err) {
            caughtError = err;
        }
        expect(caughtError).toBeInstanceOf(ImperativeError);
        expect(caughtError.message).toEqual("Unsupported data set type.");
    });

    it("should return succes:false if we failed to create the dataset", async () => {
        const connection = { submitJCL: jest.fn() };
        jest.spyOn(IO, "readFileSync").mockReturnValue(Buffer.from("//JOBCARD"));
        jest.spyOn(JobUtils, "findJobByID").mockResolvedValue("SOME ERROR" as any);

        // Dummy options to touch all `else` branches
        const options: any = { printAttributes: false, size: "1", volser: "dummy", unit: "dummy" };
        const response = await CreateDataset.create(connection, 0, "dsname", "jobCard", options);

        expect(connection.submitJCL.mock.calls[0][0]).toMatchSnapshot();
        expect(response.commandResponse).toContain("Failed to create data set");
    });

    it("should return succes:false if we failed to create the dataset", async () => {
        const connection = { submitJCL: jest.fn() };
        jest.spyOn(IO, "readFileSync").mockReturnValue(Buffer.from("//JOBCARD"));
        jest.spyOn(JobUtils, "findJobByID").mockResolvedValue("SOME ERROR" as any);

        // Dummy options to touch all `other` branches
        const options: any = { printAttributes: true, size: "1U", secondary: 1, volser: "dummy", unit: "dummy" };
        const response = await CreateDataset.create(connection, 0, "dsname", "jobCard", options);

        expect(connection.submitJCL.mock.calls[0][0]).toMatchSnapshot();
        expect(response.commandResponse).toContain("Failed to create data set");
    });

    describe("Should create datasets successfully", () => {
        const testCreate = async (type: CreateDataSetTypeEnum): Promise<boolean> => {
            const connection = { submitJCL: jest.fn() };
            jest.spyOn(IO, "readFileSync").mockReturnValue(Buffer.from("//JOBCARD"));
            jest.spyOn(JobUtils, "findJobByID").mockResolvedValue("RC=0000" as any);

            const response = await CreateDataset.create(connection, type, "dsname", "jobCard");

            expect(connection.submitJCL.mock.calls[0][0]).toMatchSnapshot();
            expect(response.commandResponse).toContain("Data set created successfully.");
            return response.success;
        }
        it("partitioned", async () => {
            const ret = await testCreate(CreateDataSetTypeEnum.DATA_SET_PARTITIONED);
            expect(ret).toBe(true);
        });

        it("sequential", async () => {
            const ret = await testCreate(CreateDataSetTypeEnum.DATA_SET_SEQUENTIAL);
            expect(ret).toBe(true);
        });

        it("binary", async () => {
            const ret = await testCreate(CreateDataSetTypeEnum.DATA_SET_BINARY);
            expect(ret).toBe(true);
        });

        it("C", async () => {
            const ret = await testCreate(CreateDataSetTypeEnum.DATA_SET_C);
            expect(ret).toBe(true);
        });

        it("classic", async () => {
            const ret = await testCreate(CreateDataSetTypeEnum.DATA_SET_CLASSIC);
            expect(ret).toBe(true);
        });
    });
});
