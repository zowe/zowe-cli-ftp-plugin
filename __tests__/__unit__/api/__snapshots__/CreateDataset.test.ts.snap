// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateDataset Should create datasets successfully C 1`] = `
"//JOBCARD
//CREATE EXEC PGM=IEFBR14
//NEWDATA DD DSN=DSNAME,DISP=(NEW,CATLG),SPACE=(CYL,(1)),LRECL=260,
//   RECFM=VB,BLKSIZE=32760,DSORG=PO"
`;

exports[`CreateDataset Should create datasets successfully binary 1`] = `
"//JOBCARD
//CREATE EXEC PGM=IEFBR14
//NEWDATA DD DSN=DSNAME,DISP=(NEW,CATLG),SPACE=(CYL,(10)),
//   LRECL=27998,RECFM=U,BLKSIZE=27998,DSORG=PO"
`;

exports[`CreateDataset Should create datasets successfully classic 1`] = `
"//JOBCARD
//CREATE EXEC PGM=IEFBR14
//NEWDATA DD DSN=DSNAME,DISP=(NEW,CATLG),SPACE=(CYL,(1)),LRECL=80,
//   RECFM=FB,BLKSIZE=6160,DSORG=PO"
`;

exports[`CreateDataset Should create datasets successfully partitioned 1`] = `
"//JOBCARD
//CREATE EXEC PGM=IEFBR14
//NEWDATA DD DSN=DSNAME,DISP=(NEW,CATLG),SPACE=(CYL,(1)),LRECL=80,
//   RECFM=FB,BLKSIZE=6160,DSORG=PO"
`;

exports[`CreateDataset Should create datasets successfully sequential 1`] = `
"//JOBCARD
//CREATE EXEC PGM=IEFBR14
//NEWDATA DD DSN=DSNAME,DISP=(NEW,CATLG),SPACE=(CYL,(1)),LRECL=80,
//   RECFM=FB,BLKSIZE=6160,DSORG=PS"
`;

exports[`CreateDataset should return succes:false if we failed to create the dataset 1`] = `
"//JOBCARD
//CREATE EXEC PGM=IEFBR14
//NEWDATA DD DSN=DSNAME,DISP=(NEW,CATLG),SPACE=(CYL,(1,0)),
//   VOLSER=DUMMY,LRECL=27998,RECFM=U,BLKSIZE=27998,UNIT=DUMMY,
//   DSORG=PO"
`;

exports[`CreateDataset should return succes:false if we failed to create the dataset with additional options 1`] = `
"//JOBCARD
//CREATE EXEC PGM=IEFBR14
//NEWDATA DD DSN=DSNAME,DISP=(NEW,CATLG),SPACE=(U,(1,1)),
//   VOLSER=DUMMY,LRECL=27998,RECFM=U,BLKSIZE=27998,UNIT=DUMMY,
//   DSORG=PO"
`;
